integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java

```diff
@@ -321,4 +321,89 @@ public void testLoanClassificationJob() {
+
+        LocalDate todaysDate = Utils.getLocalDateOfTenant();
+        todaysDate = todaysDate.minusDays(40);
+        String operationDate = Utils.dateFormatter.format(todaysDate);
+        final String principalAmount = "10000";
+
+        final String loanApplicationJSON = new LoanApplicationTestBuilder().withPrincipal(principalAmount).withLoanTermFrequency("12")
+                .withLoanTermFrequencyAsMonths().withNumberOfRepayments("12").withRepaymentEveryAfter("1")
+                .withRepaymentFrequencyTypeAsMonths() //
+                .withInterestRatePerPeriod("2") //
+                .withExpectedDisbursementDate(operationDate) //
+                .withInterestTypeAsDecliningBalance() //
+                .withSubmittedOnDate(operationDate) //
+                .build(clientId.toString(), loanProductId.toString(), null);
+        final Integer loanId = loanTransactionHelper.getLoanId(loanApplicationJSON);
+        loanTransactionHelper.approveLoan(operationDate, principalAmount, loanId, null);
+        loanTransactionHelper.disburseLoanWithNetDisbursalAmount(operationDate, loanId, principalAmount);
+
+        GetLoansLoanIdResponse getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after Disbursement {}", getLoansLoanIdResponse.getDelinquencyRange().getClassification());
+        assertNotNull(getLoansLoanIdResponse);
+        // First Loan Delinquency Classification after Disbursement command
+        assertEquals(getLoansLoanIdResponse.getDelinquencyRange().getClassification(), classificationExpected);
+
+        // When
+        // Apply a partial repayment
+        todaysDate = Utils.getLocalDateOfTenant();
+        operationDate = Utils.dateFormatter.format(todaysDate);
+        loanTransactionHelper.makeRepayment(operationDate, 100.0f, loanId);
+
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after Repayment {}", getLoansLoanIdResponse.getDelinquencyRange());
+        assertNotNull(getLoansLoanIdResponse.getDelinquencyRange());
+        // First Loan Delinquency Classification remains after Repayment because the installment is not fully paid
+        assertEquals(getLoansLoanIdResponse.getDelinquencyRange().getClassification(), classificationExpected);
+
+        // Apply a repayment to get a full paid installment
+        loanTransactionHelper.makeRepayment(operationDate, 1000.0f, loanId);
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after Repayment {}", getLoansLoanIdResponse.getDelinquencyRange());
+        assertNotNull(getLoansLoanIdResponse);
+        // The Loan Delinquency Classification after Repayment command must be null
+        assertNull(getLoansLoanIdResponse.getDelinquencyRange());
+    }
+
 }
```

Comment:
@josehernandezfintecheandomx Would you please add more test cases to cover the proper behaviour:
- Check whether the history entries are correct
- Check what happens if there are at least 2 unpaid instalment and the oldest got fully paid, but the 2nd is not
- Check what happens if a backdated repayment make the instalment unpaid
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted
- Check what happens if the loan was delinquent but it got fully repaid and closed
- Check what happens if the loan was delinquent but it got overpaid

Reply:
Done, Integration tests extended, here is an image with part of the output

<img width="1606" alt="Screen Shot 2022-09-04 at 23 39 33" src="https://user-images.githubusercontent.com/44206706/188362096-769ce8f0-138d-42f9-9f9d-4e798776b297.png">

Reply:
The following use cases are still not covered:
- Check what happens if a backdated repayment make the instalment unpaid -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the loan was delinquent but it got fully **prepaid** and closed -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the loan was delinquent but it got fully **prepaid** and closed -> will the delinquency classfication applied? History entries are correct?

Reply:
@josehernandezfintecheandomx Would you take a look please?

Reply:
I think we should have at least one test case where there is repayment reversal